### PR TITLE
Adds the ``BlockPreDispense`` event.

### PR DESCRIPTION
**Pull request by heypr**
# Additions
- The Paper-only``BlockPreDispense`` event, ``<block> predispenses <item>``.
  - ``<context.slot>`` returns a ``ElementTag(Number)`` of the slot that will be dispensed from.
  - ``<context.item>`` returns the ``ItemTag`` of the item about to be dispensed.
  - ``<context.location>`` returns the ``LocationTag`` of the dispenser.

Requested by [Yukiyahana](https://discord.com/channels/315163488085475337/1169387019798073414) on Discord.

### CODE REVIEW

File: paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    // @Events
+    // <block> tries to dispense <item>
+    //
+    // @Group Block
+    //
+    // @Location true
+    //
+    // @Cancellable true
+    //
+    // @Triggers before a block dispenses an item.
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java:24]: There should be an explanation here of when/why you would use this event vs the existing `<block> dispenses <item>`

File: paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    //
+    // @description
+    // This event fires before the dispenser finishes processing its drop, allowing access to the slot trying to be dispensed and full cancellation of the dispenser's sounds.
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java]: I don't think this is a thing? should be under `Triggers` afaik ([see](https://github.com/DenizenScript/Denizen/blob/9afda6bf726e841356b4fc0c12ce1d0c8c2f7842/plugin/src/main/java/com/denizenscript/denizen/events/entity/EntityDeathScriptEvent.java#L39-L40))

File: paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    // @Events
+    // <block> tries to dispense <item>
+    //
+    // @Group Block
+    //
+    // @Location true
+    //
+    // @Cancellable true
+    //
+    // @Triggers before a block dispenses an item. This event fires before the dispenser finishes processing its drop, allowing access to the slot trying to be dispensed and full cancellation of the dispenser's sounds.
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java]: You can have lines below it as part of the meta (e.g. [delta time event](https://github.com/DenizenScript/Denizen-Core/blob/25cf2298fcdcf282757005ec83a637595a6b106a/src/main/java/com/denizenscript/denizencore/events/core/DeltaTimeScriptEvent.java#L18-L21)), probably makes sense to put the explanation bit in a new line.

File: paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    // -->
+
+    public BlockPreDispenseScriptEvent() {
+        registerCouldMatcher("<block> tries to dispense <item>");
+    }
+
+    public BlockPreDispenseEvent event;
+    public ItemTag item;
+    public LocationTag location;
+    @Override
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/events/BlockPreDispenseScriptEvent.java:42]: Missing newline

**Pull request merged**